### PR TITLE
Dim tracking essentials

### PR DESCRIPTION
**Pull request by jane-intel**

### CODE REVIEW

File: src/core/src/dimension_tracker.hpp [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    void set_table_of_equivalence(ov::Dimension& d) const {
+        OPENVINO_ASSERT(d.m_table_of_equivalence == nullptr, "ov::Dimension is already being tracked");
+        OPENVINO_ASSERT(m_table_of_equivalence != nullptr,
+                        "Can not set nullptr as table of equivalence shared pointer");
+        d.m_table_of_equivalence = m_table_of_equivalence;
+    }
+
+    std::shared_ptr<TableOfEquivalence> get_table_of_equivalence(ov::Dimension& d) const {
```
Comment (ilyachur) [inline src/core/src/dimension_tracker.hpp:39]:
```suggestion
    const std::shared_ptr<TableOfEquivalence>& get_table_of_equivalence(ov::Dimension& d) const {
```

File: src/core/src/partial_shape.cpp [hunk 2]
```diff
@@ -130,6 +131,8 @@ std::ostream& ov::operator<<(std::ostream& str, const PartialShape& shape) {
             if (!first) {
                 str << ",";
             }
+            if (const auto l = ov::DimensionTracker::get_label(d))
```
Comment (ilyachur) [inline src/core/src/partial_shape.cpp:134]: Could we add tests?

File: src/common/transformations/include/transformations/dimension_tracking.hpp [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+public:
+    OPENVINO_RTTI("FindBatch");
+    bool run_on_model(const std::shared_ptr<ngraph::Function>& m) override;
+};
+
+namespace ov {
+class DimensionTracker;
+
+namespace batch_util {
+    std::map<ov::opset1::Parameter*, ov::PartialShape> collect_original_shapes(std::shared_ptr<ov::Model> f);
```
Comment (jane-intel) [inline src/common/transformations/include/transformations/dimension_tracking.hpp:35]: const shared_prt

File: src/core/src/pass/dimension_tracking.cpp [hunk 4]
```diff
@@ -0,0 +1,296 @@
+// Copyright (C) 2018-2021 Intel Corporation
```
Comment (ilya-lavrenov) [inline src/core/src/pass/dimension_tracking.cpp:1]: why is it a common functionality? not a part of transformation library?

File: src/tests/functional/inference_engine/CMakeLists.txt [hunk 5]
```diff
@@ -4,7 +4,8 @@
 
 set(TARGET_NAME ieFuncTests)
 
-set(INCLUDES $<TARGET_PROPERTY:inference_engine_obj,SOURCE_DIR>/src)
+set(INCLUDES $<TARGET_PROPERTY:inference_engine_obj,SOURCE_DIR>/src
+    ${OpenVINO_SOURCE_DIR}/src/core/src)
```
Comment (ilya-lavrenov) [inline src/tests/functional/inference_engine/CMakeLists.txt:8]: Use https://github.com/openvinotoolkit/openvino/pull/9571 once it's merged

Comment (ilya-lavrenov) [reply src/tests/functional/inference_engine/CMakeLists.txt:8]: Merged

File: src/common/transformations/include/transformations/dimension_tracking.hpp [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+#pragma once
+
+#include <memory>
+
+#include <transformations_visibility.hpp>
+
+#include <openvino/pass/graph_rewrite.hpp>
+#include <openvino/opsets/opset1.hpp>
+
+using namespace std;
```
[the remark this answers is not in the document]
Comment (ilyachur) [reply src/common/transformations/include/transformations/dimension_tracking.hpp]: Using std still there

File: src/common/transformations/include/transformations/dimension_tracking.hpp [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+using P2Btype = std::unordered_map<std::shared_ptr<ov::opset1::Parameter>, std::unordered_set<size_t>>;
```
Comment (ilyachur) [inline src/common/transformations/include/transformations/dimension_tracking.hpp:14]: Maybe more clear name: Parameter2BatchType?

File: src/common/transformations/include/transformations/dimension_tracking.hpp [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+class DimensionTracker;
+
+namespace batch_util {
+    void mark_batch(std::shared_ptr<ov::opset1::Parameter> &parameter, P2Btype &map, const std::unordered_set<size_t> &batches);
+    void mark_no_batch(std::shared_ptr<ov::opset1::Parameter> &parameter, P2Btype &map);
+    void mark_layout_independent_batch(std::shared_ptr<ov::opset1::Parameter> &parameter, const std::shared_ptr<ov::Node> & result, P2Btype &map);
+    void mark_with_unique_dimension_labels(const std::shared_ptr<Model> &f, const ov::DimensionTracker &dt);
+    void restore_original_dimensions_except_batch(const std::map<std::shared_ptr<ov::opset1::Parameter>, ov::PartialShape>& parameter_to_shape);
+    bool check_batch_tracks_through_all_the_nodes(const std::shared_ptr<ov::Model>& f);
+    P2Btype find_batch(const std::shared_ptr<ov::Model> &f);
```
Comment (ilyachur) [inline src/common/transformations/include/transformations/dimension_tracking.hpp:40]:
```suggestion
    P2Btype find_batch(const std::shared_ptr<ov::Model> &m);
```

File: src/common/transformations/include/transformations/dimension_tracking.hpp [hunk 9]
```diff
@@ -0,0 +1,43 @@
+// Copyright (C) 2018-2021 Intel Corporation
```
Comment (ilyachur) [inline src/common/transformations/include/transformations/dimension_tracking.hpp:1]: Why not in the `src/common/transformations/include/transformations/common_optimizations/dimension_tracking.hpp` as source file?

File: src/common/transformations/src/transformations/common_optimizations/dimension_tracking.cpp [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+        {ov::opset1::GroupConvolution::get_type_info_static(), {0, 0}},
+        {ov::opset1::ConvolutionBackpropData::get_type_info_static(), {0, 0}},
+        {ov::opset1::GroupConvolutionBackpropData::get_type_info_static(), {0, 0}},
+        {ov::opset1::DeformableConvolution::get_type_info_static(), {0, 0}},
+        {ov::opset1::MatMul::get_type_info_static(), {0, 0}},  // transpose_a situation
+    };
+
+    P2Btype parameter_to_batch_labels;
+
+    for (auto parameter : f->get_parameters()) {
```
Comment (ilyachur) [inline src/common/transformations/src/transformations/common_optimizations/dimension_tracking.cpp:101]: const auto&

File: src/common/transformations/src/transformations/common_optimizations/dimension_tracking.cpp [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+                ov::DimensionTracker::set_label(original_shape[n], label);
+            }
+        }
+        item.first->set_partial_shape(original_shape);
+    }
+}
+
+bool ov::batch_util::check_batch_tracks_through_all_the_nodes(const std::shared_ptr<ov::Model>& f) {
+    bool failed_to_propagate_batch = false;
+    for (auto& node : f->get_ordered_ops()) {
```
Comment (ilyachur) [inline src/common/transformations/src/transformations/common_optimizations/dimension_tracking.cpp:181]: const auto&

File: src/common/transformations/src/transformations/common_optimizations/dimension_tracking.cpp [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+    return failed_to_propagate_batch;
+}
+
+bool ov::pass::FindBatch::run_on_model(const std::shared_ptr<ov::Model>& f) {
+    auto te = std::make_shared<ov::TableOfEquivalence>();
+    ov::DimensionTracker dt(te);
+
+    const auto& parameters = f->get_parameters();
+    std::map<std::shared_ptr<ov::opset1::Parameter>, PartialShape> parameter_to_shape;
+    for (auto& parameter : parameters) {
```
Comment (ilyachur) [inline src/common/transformations/src/transformations/common_optimizations/dimension_tracking.cpp:229]: const auto&

File: src/tests/functional/inference_engine/CMakeLists.txt [hunk 13]
```diff
@@ -11,16 +11,15 @@ set(LINK_LIBRARIES
...
     openvino::itt
     openvino::util
     openvino::conditional_compilation
     sharedTestClasses
     inference_engine_snippets
     offline_transformations
-    inference_engine
-    frontend_common
     pugixml::static
+    inference_engine
```
Comment (ilyachur) [inline src/tests/functional/inference_engine/CMakeLists.txt:21]:
```suggestion
```

**Review by ilyachur — APPROVED** [review]
Comments can be fixed in next PR

File: src/tests/functional/inference_engine/transformations/dimension_tracking.cpp [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+    const auto& f = std::make_shared<ov::Model>(ov::NodeVector{conv_0, conv_1}, ov::ParameterVector{data});
+
+    ov::pass::Manager m;
+    m.register_pass<ngraph::pass::InitNodeInfo>();
+    m.register_pass<ov::pass::FindBatch>();
+    m.run_passes(f);
+    ASSERT_NO_THROW(check_rt_info(f));
+
+    const auto& shape = data->get_partial_shape();
+    ASSERT_TRUE(ov::DimensionTracker::get_label(shape[0])) << shape;
```
Comment (myshevts) [inline src/tests/functional/inference_engine/transformations/dimension_tracking.cpp:201]: pls add/update tests (in another PR) for **outputs** labelling

**Pull request merged**